lib/gemstash/gem_source/rack_middleware.rb

```diff
@@ -0,0 +1,22 @@
+require "gemstash"
+
+module Gemstash
+  module GemSource
+    # Rack middleware to detect the gem source from the URL.
+    class RackMiddleware
+      def initialize(app)
+        @app = app
+      end
+
+      def call(env)
+        Gemstash::GemSource.sources.each do |source|
+          next unless source.matches?(env)
+          env["gemstash.gem_source"] = source
+          break
+        end
```

Comment:
I like the idea of capturing the upstream in the middleware, that way all the methods are affected yet not changed.